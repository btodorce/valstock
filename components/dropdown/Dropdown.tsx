import React, { FC, memo, ReactNode, useState } from "react";
import useMeasure from "react-use-measure";
import { useSpring, a } from "@react-spring/web";
import styles from "./index.module.scss";
import { ResizeObserver } from "@juggle/resize-observer";

export interface P {
    title: string;
    children: ReactNode;
}

export const Dropdown: FC<P> = ({ title, children }) => {
    const [isOpened, setIsOpened] = useState(false);
    const [ref, { height: viewHeight }] = useMeasure({
        polyfill: ResizeObserver,
    });

    const animProps = useSpring({
        height: isOpened ? viewHeight : 0,
        config: { tension: 250, friction: 32, clamp: true, duration: 150 },
        opacity: isOpened ? 1 : 0,
    });

    const handleClickEvent = () => setIsOpened((x) => !x);

    return (
        <div
            className={styles.root}
            role="button"
            tabIndex={0}
            aria-expanded={isOpened}
            onClick={handleClickEvent}
        >
            <div className={styles.header}>
                <span className={styles.label}>{title}</span>
            </div>
            <a.div style={{ overflow: "hidden", ...animProps }}>
                <div ref={ref} className={styles.content}>
                    {children}
                </div>
            </a.div>
        </div>
    );
};

Dropdown.displayName = "Dropdown";
export default memo(Dropdown);
