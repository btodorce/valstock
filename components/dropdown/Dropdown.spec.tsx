import { render } from "@testing-library/react";
import renderer from "react-test-renderer";
import Dropdown from "./Dropdown";

const defaultProps = {
    title: "Dropdown title",
    children: (
        <ul>
            <li>Test</li>
            <li>Test 2</li>
        </ul>
    )
};

describe("Button", () => {
    it("should render successfully", () => {
        const tree = renderer.create(<Dropdown {...defaultProps} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
