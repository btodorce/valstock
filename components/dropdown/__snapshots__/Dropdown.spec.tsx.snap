// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render successfully 1`] = `
<div
  aria-expanded={false}
  className="root"
  onClick={[Function]}
  role="button"
  tabIndex={0}
>
  <div
    className="header"
  >
    <span
      className="label"
    >
      Dropdown title
    </span>
  </div>
  <div
    style={
      Object {
        "height": 0,
        "opacity": 0,
        "overflow": "hidden",
      }
    }
  >
    <div
      className="content"
    >
      <ul>
        <li>
          Test
        </li>
        <li>
          Test 2
        </li>
      </ul>
    </div>
  </div>
</div>
`;
